src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2825,6 +2838,133 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
             SetResult(node, GetReturnTypeWithState(method), method.ReturnTypeWithAnnotations);
         }
 
+        private bool IsStaticEqualsMethod(MethodSymbol method)
+        {
+            if (method.Equals(compilation.GetSpecialTypeMember(SpecialMember.System_Object__EqualsObjectObject)))
+            {
+                return true;
+            }
+            else if (method.Equals(compilation.GetSpecialTypeMember(SpecialMember.System_Object__ReferenceEquals)))
+            {
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        private bool IsObjectEqualsMethodOrOverride(MethodSymbol method)
+        {
+            if (method.ParameterCount != 1
+                || method.ReturnType.SpecialType != SpecialType.System_Boolean
+                || method.Name != WellKnownMemberNames.ObjectEquals)
+            {
+                return false;
+            }
+
+            var objectEqualsMethod = compilation.GetSpecialTypeMember(SpecialMember.System_Object__Equals);
+            return method.GetLeastOverriddenMethod(accessingTypeOpt: null).Equals(objectEqualsMethod);
+        }
+
+
+        private bool IsIEquatableEqualsMethodOrImplementation(MethodSymbol method)
+        {
+            if (method.ParameterCount != 1
+                || method.MethodKind != MethodKind.Ordinary
+                || method.ReturnType.SpecialType != SpecialType.System_Boolean
+                || method.Name != WellKnownMemberNames.ObjectEquals)
+            {
+                return false;
+            }
+
+            var iEquatableType = compilation.GetWellKnownType(WellKnownType.System_IEquatable_T);
+            if (iEquatableType is null)
+            {
+                return false;
+            }
+
+            var parameterType = method.Parameters[0].TypeWithAnnotations;
+            var constructedIEquatableType = iEquatableType.Construct(ImmutableArray.Create(parameterType));
+            var members = constructedIEquatableType.GetMembers(WellKnownMemberNames.ObjectEquals);
+            if (members.Length != 1)
+            {
+                // If the Equals method is missing on the interface definition, or it's overloaded, give up.
```

Comment:
>or it's overloaded, give up. [](start = 80, length = 28)

I don't imagine that the BCL is going to add an `Equals` overload on `object`, but it's technically possible for one to be there. Can we pick the one with the `object` parameter type?

Reply:
The more "common" scenario is that `bool IEquatable<T>.Equals(T)` is not defined at all, but this is really just a "degrade gracefully in the face of bad metadata" behavior.

re: `GetWellKnownTypeMember`, I would like to, however I am stumped as to how to actually make the comparison between the method inside the original definition (`IEquatable<T>.Equals(T)`) and the method inside a constructed symbol (`IEquatable<string?>.Equals(string?)`)

---
In reply to: [297381621](https://github.com/dotnet/roslyn/pull/36722#discussion_r297381621) [](ancestors = 297381621)